crates/agglayer-aggregator-notifier/src/certifier/mod.rs

```diff
@@ -0,0 +1,175 @@
+use std::sync::Arc;
+
+use agglayer_certificate_orchestrator::{Certifier, CertifierOutput, Error};
+use agglayer_prover_types::{
+    default_bincode_options,
+    v1::{
+        proof_generation_service_client::ProofGenerationServiceClient, ProofGenerationRequest,
+        ProofGenerationResponse,
+    },
+};
+use agglayer_storage::stores::{PendingCertificateReader, PendingCertificateWriter};
+use agglayer_types::{Height, LocalNetworkStateData, NetworkId, Proof};
+use bincode::Options as _;
+use futures::future::BoxFuture;
+use pessimistic_proof::{generate_pessimistic_proof, LocalNetworkState};
+use reth_primitives::Address;
+use sp1_sdk::{CpuProver, Prover as _, SP1VerifyingKey};
+use tonic::transport::Channel;
+use tracing::{debug, error, info, warn};
+
+use crate::ELF;
+
+#[derive(Clone)]
+pub struct CertifierClient<PendingStore> {
+    pending_store: Arc<PendingStore>,
+    prover: ProofGenerationServiceClient<Channel>,
+    verifier: Arc<CpuProver>,
+    verifying_key: SP1VerifyingKey,
+}
+
+impl<PendingStore> CertifierClient<PendingStore> {
+    pub async fn try_new(prover: String, pending_store: Arc<PendingStore>) -> anyhow::Result<Self> {
+        let verifier = CpuProver::new();
+        let (_, verifying_key) = verifier.setup(ELF);
+
+        Ok(Self {
+            pending_store,
+            prover: ProofGenerationServiceClient::connect(prover).await?,
+            verifier: Arc::new(verifier),
+            verifying_key,
+        })
+    }
+}
+
+impl<PendingStore> Certifier for CertifierClient<PendingStore>
+where
+    PendingStore: PendingCertificateReader + PendingCertificateWriter + 'static,
+{
+    fn certify(
+        &self,
+        mut state: LocalNetworkStateData,
+        network_id: NetworkId,
+        height: Height,
+    ) -> Result<BoxFuture<Result<CertifierOutput, Error>>, Error> {
+        debug!("Certifying the certificate of network {network_id} at height {height}");
+
+        // Fetch certificate from storage
+        let certificate = self
+            .pending_store
+            .get_certificate(network_id, height)?
+            .ok_or(Error::CertificateNotFound(network_id, height))?;
+
+        let certificate_id = certificate.hash();
+
+        if self.pending_store.get_proof(certificate_id)?.is_some() {
+            return Err(Error::ProofAlreadyExists(network_id, height));
+        }
+
+        let signer = Address::new([0; 20]); // TODO: put the trusted sequencer address
+
+        let initial_state = LocalNetworkState::from(state.clone());
+        let multi_batch_header =
+            state
+                .apply_certificate(&certificate, signer)
+                .map_err(|source| Error::Types {
+                    certificate_id,
+                    source,
+                })?;
+
+        // Perform the native PP execution
+        let _ = generate_pessimistic_proof(initial_state.clone(), &multi_batch_header).map_err(
```

Comment:
Nitpick: Would it make sense to change `generate_pessimistic_proof` to accept the initial state by ref? Wold avoid the clone here.

Reply:
I think that the `initial_network_state` need to be own in order to apply the certificate on it, if it fails we do not want to touch the original `network_state`